felix/fv/pre_dnat_test.go

```diff
@@ -200,6 +206,87 @@ var _ = infrastructure.DatastoreDescribe("pre-dnat with initialized Felix, 2 wor
 				})
 
 				Context("with pre-DNAT policy to open pinhole to 8055", func() {
+					// Exec iptables-save, or iptables-save -t [table], if table is not "".
+					dumpIPTables := func(table string) []string {
+						execArgs := []string{"iptables-save"}
+						if table != "" {
+							execArgs = append(execArgs, "-t", table)
+						}
+						out, err := tc.Felixes[0].ExecOutput(execArgs...)
+						Expect(err).NotTo(HaveOccurred())
+						lines := strings.Split(out, "\n")
+						Expect(len(lines)).NotTo(Equal(1), "couldn't split on newline")
+						var grep []string
+						for _, line := range lines {
+							if strings.Contains(line, "cali-") && strings.Contains(line, "-A ") {
+								grep = append(grep, line)
+							}
+						}
+						return grep
+					}
+					dumpMangleTable := func() []string { return dumpIPTables("mangle") }
+					dumpFilterTable := func() []string { return dumpIPTables("filter") }
+
+					// Returns nil when the value of f() hasn't changed for the duration settledPeriod.
+					// Returns error if f() never settles up to the timeout.
+					// Immediately returns error if timeout is lower than settledPeriod.
+					waitToSettle := func(f func() int, settledPeriod, timeout time.Duration) error {
```

Comment:
As discussed on slack, would prefer not to rely on settling, if there's something explicit that we can wait for.  It's not a bad approach, it's just a bit slow (and if it flakes, you have to bump the settled timeout so it gets slower...)

If we do go down this path, I suggest making it a general utility method.  Kubernetes has a `wait` package with similar utility funcs in it.